upup/pkg/fi/cloudup/awsup/aws_verifier.go

```diff
@@ -0,0 +1,149 @@
+	"encoding/json"
+	"encoding/xml"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/sts"
+	"k8s.io/kops/upup/pkg/fi"
+)
+
+type awsVerifier struct {
+	sts    *sts.STS
+	client http.Client
+}
+
+var _ fi.Verifier = &awsVerifier{}
+
+func NewAWSVerifier() (fi.Verifier, error) {
+	config := aws.NewConfig().WithCredentialsChainVerboseErrors(true)
+	sess, err := session.NewSession(config)
+	if err != nil {
+		return nil, err
+	}
+	return &awsVerifier{
+		sts: sts.New(sess),
+		client: http.Client{
+			Transport: &http.Transport{
+				Proxy: http.ProxyFromEnvironment,
+				DialContext: (&net.Dialer{
+					Timeout: 30 * time.Second,
+				}).DialContext,
+				ForceAttemptHTTP2:     true,
+				DisableKeepAlives:     true,
+				MaxIdleConnsPerHost:   -1,
+				TLSHandshakeTimeout:   10 * time.Second,
+				ExpectContinueTimeout: 1 * time.Second,
+			},
+			CheckRedirect: func(req *http.Request, via []*http.Request) error {
+				return http.ErrUseLastResponse
+			},
+		},
+	}, nil
+}
+
+type GetCallerIdentityResponse struct {
+	XMLName                 xml.Name                  `xml:"GetCallerIdentityResponse"`
+	GetCallerIdentityResult []GetCallerIdentityResult `xml:"GetCallerIdentityResult"`
+	ResponseMetadata        []ResponseMetadata        `xml:"ResponseMetadata"`
+}
+
+type GetCallerIdentityResult struct {
+	Arn     string `xml:"Arn"`
+	UserId  string `xml:"UserId"`
+	Account string `xml:"Account"`
+}
+
+type ResponseMetadata struct {
+	RequestId string `xml:"RequestId"`
+}
+
+func (a awsVerifier) VerifyToken(token string, body []byte) (string, error) {
+	if !strings.HasPrefix(token, AWSAuthenticationTokenPrefix) {
+		return "", fmt.Errorf("incorrect authorization type")
+	}
+	token = strings.TrimPrefix(token, AWSAuthenticationTokenPrefix)
+
+	// We rely on the client and server using the same version of the same STS library.
+	stsRequest, _ := a.sts.GetCallerIdentityRequest(nil)
+	err := stsRequest.Sign()
+	if err != nil {
+		return "", fmt.Errorf("creating identity request: %v", err)
+	}
+
+	stsRequest.HTTPRequest.Header = nil
+	tokenBytes, err := base64.StdEncoding.DecodeString(token)
+	if err != nil {
+		return "", fmt.Errorf("decoding authorization token: %v", err)
+	}
+	err = json.Unmarshal(tokenBytes, &stsRequest.HTTPRequest.Header)
+	if err != nil {
+		return "", fmt.Errorf("unmarshalling authorization token: %v", err)
+	}
+
+	sha := sha256.Sum256(body)
+	if stsRequest.HTTPRequest.Header.Get("X-Kops-Request-SHA") != base64.RawStdEncoding.EncodeToString(sha[:]) {
+		return "", fmt.Errorf("incorrect SHA")
+	}
+
+	requestBytes, _ := ioutil.ReadAll(stsRequest.Body)
+	_, _ = stsRequest.Body.Seek(0, io.SeekStart)
+	if stsRequest.HTTPRequest.Header.Get("Content-Length") != strconv.Itoa(len(requestBytes)) {
+		return "", fmt.Errorf("incorrect content-length")
+	}
+
+	// TODO - implement retry?
```

Comment:
I suspect our client needs to retry anyway, so there might not be much reason for us to retry here.

Reply:
Nodeup will retry, it's just a question of how long transient errors delay node startup.